### PR TITLE
created users endpoint to post and get

### PR DESCRIPTION
**Pull request by mjmandelah07**
Endpoints for post and get users

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,25 +1,35 @@
...
-  try {
-    // await db.connect();
-    await createUsersAccount();
-  } catch (err) {
-    console.error(err);
-  }
-};
-database();
+//  routes
+app.use("/api/users", userRoutes);
```
Comment (kelvinator07) [inline index.js]: Read The PDF For The API Specifications

File: index.js [hunk 2]
```diff
@@ -1,25 +1,35 @@
...
+// // connect to the database
+// const database = async () => {
+//   try {
+//     // await db.connect();
+//     await createUsersAccount();
+//   } catch (err) {
+//     console.error(err);
+//   }
+// };
+// database();
```
Comment (kelvinator07) [inline index.js]: What Is This Doing Here?

File: routes/userRoutes
Comment (kelvinator07) [inline routes/userRoutes:1]: This You Test This Endpoint?
This Is Not a JavaScript File

Comment (mjmandelah07) [reply routes/userRoutes:1]: i dont know how to test yet

File: auth/auth-middle-ware
Comment (kelvinator07) [inline auth/auth-middle-ware:1]: This Is Not a JS File

Comment (mjmandelah07) [reply auth/auth-middle-ware:1]: have made changes

File: controller/article-controller.js [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+
+    res.status(200).json({
+      status: "success",
+      data: {
+        message: "Article successfully updated",
+        articleId: articleId,
+        updatedOn: createdOn,
+        title: updateResult.rows[0].title,
+        article: updateResult.rows[0].article,
+        updated_on: updateResult.rows[0].updated_on,
```
Comment (kelvinator07) [inline controller/article-controller.js]:
```suggestion
    const updatedArticle = updateResult.rows[0];

    res.status(200).json({
      status: "success",
      data: {
        message: "Article successfully updated",
        articleId: articleId,
        updatedOn: updatedArticle.created_on,
        title: updatedArticle.title,
        article: updatedArticle.article,
        updated_on: updatedArticle.updated_on,
```

File: controller/article-controller.js [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+  const user = req.user;
+  const articleData = req.body;
+
+  // Any authenticated user can create articles
+  if (!user) {
+    return res.status(401).json({
+      status: "error",
+      error: "Access denied",
+    });
+  }
```
Comment (kelvinator07) [inline controller/article-controller.js]: If The User Isn't Valid , Return From The Middleware
Don't Pass Control To The Main Function

File: controller/article-controller.js [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+  } catch (error) {
+    console.error("Error updating article:", error);
+    res.status(500).json({
+      status: "error",
+      error: "An error occurred while updating article",
+    });
+  }
+};
+
+const deleteArticleById = async (req, res) => {
```
Comment (kelvinator07) [inline controller/article-controller.js:95]: An Admin Should Also Be Able To Delete An Article

File: controller/article-controller.js [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+  }
+};
+module.exports = {
+  createArticle,
+  updateArticlebyId,
+  deleteArticleById,
+  getArticleById,
+  getAllArticlesByUserId,
+  getAllArticles,
+};
```
Comment (kelvinator07) [inline controller/article-controller.js:251]: Always Leave An Extra Line In Each File

File: controller/article-controller.js (hunk 3 above)
Comment (mjmandelah07) [reply controller/article-controller.js]: fixed

File: controller/article-controller.js (hunk 4 above)
Comment (mjmandelah07) [reply controller/article-controller.js]: fixed

File: controller/article-controller.js (hunk 6 above)
Comment (mjmandelah07) [reply controller/article-controller.js:251]: fixed

File: controller/gif-controller.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  // get the token details from the authmiddleware
+  const user = req.user;
+  const { base64EncodedGif, title } = req.body;
+
+  // configure cloudinary
+  cloudinary.config({
+    cloud_name: process.env.CLOUD_NAME,
+    api_key: process.env.CLOUD_API_KEY,
+    api_secret: process.env.CLOUD_API_KEY_SECRET,
+  });
```
Comment (kelvinator07) [inline controller/gif-controller.js]: Do This Configuration In Another File , Then Import Here

File: controller/gif-controller.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    res.status(500).json({
+      status: "error",
+      error: "An error occurred while creating gif",
+    });
+  }
+};
+
+const getAllGifs = async (req, res) => {
+  try {
+    const selectQuery = "SELECT id, url, title, user_id, created_on FROM gifs";
```
Comment (kelvinator07) [inline controller/gif-controller.js]:
```suggestion
    const selectQuery = "SELECT * FROM gifs";
```

File: controller/gif-controller.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
+const db = require("../db/db");
+const cloudinary = require("cloudinary");
+require("dotenv").config();
+
+const createGif = async (req, res) => {
+  // get the token details from the authmiddleware
+  const user = req.user;
```
Comment (kelvinator07) [inline controller/gif-controller.js]:
```suggestion
  const userId = req.user?.id;
```
Read About [Optional Chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)

File: controller/gif-controller.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    const createTableQuery = `
+          CREATE TABLE IF NOT EXISTS gifs (
+            id SERIAL PRIMARY KEY,
+            url TEXT,
+            title TEXT,
+            user_id INT,
+            created_on TIMESTAMP DEFAULT NOW()
+          );
+        `;
+    await db.query(createTableQuery);
```
Comment (kelvinator07) [inline controller/gif-controller.js]: Why Are We Doing This Here?

File: controller/gif-controller.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+    // Insert the URL from cloudinary, title, and user ID into the gifs table
+    const insertQuery = `
+    INSERT INTO gifs (url, title, user_id)
+    VALUES ($1, $2, $3) RETURNING id, created_on;
+    `;
+
+    const insertValues = [uploadResult.secure_url, title, userId];
```
Comment (kelvinator07) [inline controller/gif-controller.js]:
```suggestion
   const uploadedImageUrl = uploadResult.secure_url;
    const insertValues = [uploadedImageUrl, title, userId];
```

File: controller/gif-controller.js [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+
+    // Send the response if successful
+    res.status(201).json({
+      status: "success",
+      data: {
+        gifId: gifId,
+        message: "GIF image successfully posted",
+        createdOn: createdOn,
+        title: title,
+        imageUrl: uploadResult.secure_url,
```
Comment (kelvinator07) [inline controller/gif-controller.js]:
```suggestion
        imageUrl: uploadedImageUrl,
```

File: controller/user-controller.js [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+    const tokenPayload = {
+      userId: createdUser.id,
+      email: createdUser.email,
+      role: createdUser.role,
+    };
+
+    // Sign the payload with the secret key and generate the token
+    const token = jwt.sign(tokenPayload, process.env.JWT_SECRET, {
+      expiresIn: "24h",
+    });
```
Comment (kelvinator07) [inline controller/user-controller.js]: You Don't Need a Token After Creating a User, We Only Need It After Login

File: db/db.js [hunk 14]
```diff
@@ -1,31 +1,26 @@
...
+const { Pool } = require("pg");
 
 const dbConfig = {
   user: process.env.DB_USER,
-  host: 'localhost',
+  host: "localhost",
   database: process.env.DB_DATABASE,
   password: process.env.DB_PASSWORD,
-  port: 5432, 
+  port: 5432,
```
Comment (kelvinator07) [inline db/db.js]: Put The Host And Port In .env

File: routes/user-routes.js [hunk 15]
```diff
@@ -0,0 +1,13 @@
+const express = require("express");
+
+const { createUser, getAllUsers } = require("../controller/user-controller");
+
+const router = express.Router();
+const authMiddleware = require("../auth/auth-middle-ware");
+
+// Get all users
+router.get("/get-users", getAllUsers);
```
Comment (kelvinator07) [inline routes/user-routes.js]: Remember Only Admin Can Get All Users

File: controller/gif-controller.js (hunk 7 above)
Comment (mjmandelah07) [reply controller/gif-controller.js]: done

File: controller/article-controller.js (hunk 5 above)
Comment (mjmandelah07) [reply controller/article-controller.js:95]: done

File: controller/gif-controller.js (hunk 8 above)
Comment (mjmandelah07) [reply controller/gif-controller.js]: done

File: controller/gif-controller.js (hunk 10 above)
Comment (mjmandelah07) [reply controller/gif-controller.js]: have set another file to create dummy data

File: controller/gif-controller.js (hunk 12 above)
Comment (mjmandelah07) [reply controller/gif-controller.js]: fixed

File: controller/user-controller.js (hunk 13 above)
Comment (mjmandelah07) [reply controller/user-controller.js]: yes but the response in the pdf wants me to return token

Comment (mjmandelah07) [reply controller/user-controller.js]: after user is created

Comment (mjmandelah07) [reply controller/user-controller.js]: Have removed it thou

File: db/db.js (hunk 14 above)
Comment (mjmandelah07) [reply db/db.js]: DONE

File: routes/user-routes.js (hunk 15 above)
Comment (mjmandelah07) [reply routes/user-routes.js]: done

**Review by kelvinator07 — APPROVED** [review]

**Pull request merged**